src/porting_android.cpp

```diff
@@ -213,115 +206,83 @@ void initializePathsAndroid()
 	jmethodID mt_getAbsPath = jnienv->GetMethodID(cls_File,
 				"getAbsolutePath", "()Ljava/lang/String;");
 
-	path_cache   = getAndroidPath(nativeActivity, app_global->activity->clazz,
-			cls_File, mt_getAbsPath, "getCacheDir");
-	path_storage = getAndroidPath(cls_Env, NULL, cls_File, mt_getAbsPath,
-			"getExternalStorageDirectory");
-	path_user    = path_storage + DIR_DELIM + PROJECT_NAME_C;
-	path_share   = path_storage + DIR_DELIM + PROJECT_NAME_C;
+	path_share = getAndroidPath(nativeActivity, app_global->activity->clazz,
+			cls_File, mt_getAbsPath, "getGameSharePath");
+	path_user = getAndroidPath(nativeActivity, app_global->activity->clazz,
+			cls_File, mt_getAbsPath, "getGameUserPath");
+	path_cache = getAndroidPath(nativeActivity, app_global->activity->clazz,
+			cls_File, mt_getAbsPath, "getGameCachePath");
 
 	migrateCachePath();
 }
 
+
 void showInputDialog(const std::string& acceptButton, const  std::string& hint,
-		const std::string& current, int editType)
+		const std::string& current, AndroidEditType editType,
+		AndroidEditCallback cb, void *cb_data)
```

Comment:
std::bind would be a better way of passng data like `void *cb_data` through, IMO

Removed need to void* / static_cast, and allows more than on meta parameter
Saying that, I don't really see the need of padding this when the callback handler can get GUIFormspec as it is GUIFormspec

http://www.cplusplus.com/reference/functional/bind/